crates/ruff_server/src/server.rs

```diff
@@ -60,6 +61,42 @@ impl Server {
 
         conn.initialize_finish(id, initialize_data)?;
 
+        // Register capabilities
+        conn.sender
+            .send(lsp_server::Message::Request(lsp_server::Request {
+                id: 1.into(),
+                method: "client/registerCapability".into(),
+                params: serde_json::to_value(lsp_types::RegistrationParams {
+                    registrations: vec![lsp_types::Registration {
+                        id: "ruff-server-watch".into(),
+                        method: "workspace/didChangeWatchedFiles".into(),
+                        register_options: Some(serde_json::to_value(
+                            DidChangeWatchedFilesRegistrationOptions {
+                                watchers: vec![
+                                    FileSystemWatcher {
+                                        glob_pattern: types::GlobPattern::String(
+                                            "**/.?ruff.toml".into(),
+                                        ),
+                                        kind: None,
+                                    },
+                                    FileSystemWatcher {
+                                        glob_pattern: types::GlobPattern::String(
+                                            "**/pyproject.toml".into(),
+                                        ),
+                                        kind: None,
+                                    },
+                                ],
+                            },
+                        )?),
+                    }],
+                })?,
+            }))?;
+
+        // Flush response from server (to avoid an unexpected response appearing in the event loop)
+        let _ = conn.receiver.recv_timeout(Duration::from_secs(5)).map_err(|_| {
+            tracing::error!("Timed out while waiting for client to acknowledge registration of dynamic capabilities");
+        });
```

Comment:
Do you have ideas on how you want to support this in the future for other requests from the server that possibly happen inside of the event loop (where it isn't guaranteed that the client won't send any other requests).

Reply:
Responses from the client are received the same way as requests/notifications, so we'd handle them as part of the event loop - the scheduler would dispatch 'callback' tasks if the response is significant.